src/Datadog.Trace.ClrProfiler.Managed/Instrumentation.cs

```diff
@@ -59,7 +60,10 @@ public static void Initialize()
 
             try
             {
-                // ensure global instance is created if it's not already
+                // Creates GlobalSettings instance and loads plugins
+                PluginManager.TryLoadPlugins(GlobalSettings.Source.PluginsConfiguration);
```

Comment:
**OPTION 1**
Can we move this plugin loading inside the tracer constructor?
It looks like the constructor of the Tracer is currently the only reasonable [composition root](https://blog.ploeh.dk/2011/07/28/CompositionRoot/) for the managed code as we cannot inject anything inside Tracer.

I have also other reasons why I would advise you to do so.
Access to tracer + this Instrumentation class has synchronization mechanisms, while the `PluginManager` does not. It is very hard to reason if accessing `PluginManager.Loaded` is threat safe. I think it is but I am not sure and trying to figure it out took me quite a lot of time. Also, we may easily break the thread saftiness during refactoring (assuming everything is OK currently). The other thing I do not like is that the `PluginManager` becomes another singleton.

**OPTION 2**
We could also try changing the `Tracer.Instance` design to not use the default constructor for initialization. And pass the loaded plugins to the constructor. 

I fear that this change might be a lot harder and it may be a minefield.

Reply:
Trying out the option 2 with help of `ManualResetEventSlim` class.

Reply:
Unfortunately had to revert it. There are loads of tests and classes that hide `Tracer.Instance` calls that also expect it to create the tracer (only in case of tests). Originally the idea seem to work (running in lab).

Reply:
My only idea of how to refactor it to Option 2 without changing loads of tests would be to do something like:

```
        private static Tracer _instance = new Tracer();
        private static ManualResetEventSlim _globalInstanceResetEvent = new ManualResetEventSlim(true);
        private static object _globalInstanceLock = new object();

        /// <summary>
        /// Gets or sets the global <see cref="Tracer"/> instance.
        /// Used by all automatic instrumentation.
        /// </summary>
        public static Tracer Instance
        {
            get
            {
                _globalInstanceResetEvent.Wait();
                return _instance;
            }

            set
            {
                
                lock (_globalInstanceLock)
                {
                    _globalInstanceResetEvent.Reset();
                    _instance = value;
                    _globalInstanceResetEvent.Set();
                }
            }
        }
```

then in `Instrumentation.Initilize()` could set the `Tracer.Instance = new Tracer(plugins)` instead of hacky `_ = Tracer.Instance`

However `private static Tracer _instance = new Tracer();` would make the initialization time of the tracer about 2x longer. Also, I imagine there could be other issues as the AutoInstrumentation may start using the initial version of the tracer before it is swapped in the `Instrumentation.Initilize` method. I would personally be afraid of merging something like that 😄 

I propose to proceed with **OPTION 1** and create an issue to try to refactor to **OPTION 2** as it would require changes in a lot of tests that you mentioned.

I still think it is worth refactoring to **OPTION 1**, because:
1. We would get rid of a static class, properties, fields in `PluginManager`
2. For sure it would be thread-safe
3. The code would be covered by tests. Even if it would be a NOP or load an example plugin - at least one scenario could be "covered".

Therefore I propose to proceed with **OPTION 1** and we can create

Reply:
It seems to work more close to the original way, but the original was more performant. Plus in case of race condition you could have 2 different outcomes - it either loads Tracer with plugins or fallback to default constructor. 

We need to separate initialization and getter. The initialization needs to happen in controlled way and all other places (in case of race condition) need to wait for initialization. When its 2 in 1, it's impossible to make wrong places to wait for initialization. 

@zacharycmontoya any docs about high level execution tree from native to managed level would be highly useful.

Reply:
> Plus in case of race condition you could have 2 different outcomes - it either loads Tracer with plugins or fallback to default constructor.

If the race conditions is possible then even the current codebase may not work properly.

I would say that the approach I proposed is even safer as if the race condition would happen the plugins would eventually loaded. In case of your PR they would never be loaded + possible race condition when accessing `PluginManager.Loaded`.

Reply:
I mean the race between `Tracer.Instance` and `Tracer.Initialize()`, Otherwise I do not see any point to change LazyInitializer, as the result won't change?

I might have to try again, otherwise things won't get better. I might have to add `Tracer.Initialize()` to a few test classes constructor and this should do the trick.

Reply:
This resolves also possible `PluginManager.Loaded` issue cause plugins will be loaded before controlled initialization `Tracer.Initialize()`.  `Tracer.Instance` can't ensure controlled initialization.

Reply:
I do NOT suggest to add Tracer.Initialize() but  assign via Tracer.Instance in Instrumentation.Initialize. According to @zacharycmontoya there should be no race in case of production usage

Reply:
Yep, that should work also, so less synchronization is needed. Good point 👍